springboot/src/main/java/ca/jrvs/apps/trading/service/TraderAccountService.java

```diff
@@ -0,0 +1,188 @@
+
+  private TraderDao traderDao;
+  private AccountDao accountDao;
+  private PositionDao positionDao;
+  private SecurityOrderDao securityOrderDao;
+
+  private DecimalFormat df;
+
+  @Autowired
+  public TraderAccountService(TraderDao traderDao, AccountDao accountDao,
+      PositionDao positionDao, SecurityOrderDao securityOrderDao) {
+    this.traderDao = traderDao;
+    this.accountDao = accountDao;
+    this.positionDao = positionDao;
+    this.securityOrderDao = securityOrderDao;
+  }
+
+  /**
+   * Create a new trader and initialize a new account with 0 amount. - validate user input (all
+   * fields must be non empty) - create a trader - create an amount create, setup, and return a new
+   * traderAccountView
+   * <p>
+   * Assumption: to simplify the logic, each trader has only one account where traderId ==
+   * accountId
+   *
+   * @param trader cannot be null. All fields cannot be null except for id (auto-generated by db)
+   * @return traderAccountView
+   * @throws IllegalArgumentException if a trader has null fields or ids is not null
+   */
+  public TraderAccountView createTraderAndAccount(Trader trader) {
+
+    trader = traderExceptionCheck(trader);
+
+    Trader savedTrader = traderDao.save(trader);
+    Account newAccount = new Account();
+    newAccount.setTrader_id(savedTrader.getId());
+    newAccount.setAmount(0.0d);
+    Account savedAccount = accountDao.save(newAccount);
+
+    return new TraderAccountView(savedTrader, savedAccount);
+  }
+
+  /**
+   * Check exceptions helper function
+   */
+  public Trader traderExceptionCheck(Trader trader) {
+    if (trader == null) {
+      throw new IllegalArgumentException("trader is null");
+    }
+    if (trader.getId() != null) {
+      trader.setId(null);
+    }
+
+    for (Field field : Trader.class.getDeclaredFields()) {
+      if (field.getName().equals("id")) {
+        continue;
+      }
+      try {
+        if (field.get(field) == null) {
+          throw new IllegalArgumentException("Field: " + field.getName() + " is null");
+        }
+      } catch (IllegalAccessException e) {
+        logger.debug("Access exception when getting fields of trader", e);
+      }
+    }
+    return trader;
+  }
+
+  /**
+   * A trader can be deleted iff it has no open position and 0 cash balance - validate traderID -
+   * get trader account by traderId and check account balance - get positions by accountId and check
+   * positions - delete all securityOrders, account, trader (in this order)
+   *
+   * @param traderId must not be null
+   * @throws IllegalArgumentException if traderId is null or not found or unable to delete
+   */
+  public void deleteTraderById(Integer traderId) {
+
+    transactionSetup(traderId, 0.0d);
+
+    List<Account> returnedAccounts = accountDao.findById(traderId, true);
+
+    for (Account account : returnedAccounts) {
+      if (account.getAmount() > 0.0) {
+        throw new IllegalArgumentException(
```

Comment:
IllegalStateException fits better here. The Account is not in a delete-able state rather than the ID being unusable